res/sections/04-GameBank.tex

```diff
@@ -0,0 +1,189 @@
+\texttt{MEMBER\_CONNECTED}. Since the connection of a new member needs to be 
+broadcasted to everyone, the host forwards to all the other clients already in 
+the lobby the \texttt{BTBundle} the new member initially sent, acting in a 
+transparent way. When this phase is completed, the host informs the new client 
+with all the informations regarding the match, as the lobby name and a list of 
+players already connected. We labelled this event as \texttt{CURRENT\_STATE}. In 
+case the match comes from a previous save, the new client receives, in the same 
+event, all the transactions performed before.
+
+\begin{figure}[t]
+ \centering
+ \includegraphics[scale=0.42]{ConnectionSchema}
+ \caption{Schema of client-host connection}
+ \label{fig:gbConnectionSchema}
+\end{figure}
+
+A match can start when all the members are ready. In order to do so, users have 
+to check the option in the lobby. When this action is performed, a 
+\texttt{MEMBER\_READY} event is triggered. This event is broadcasted to everyone 
+to allow the other clients to update their UI.
+
+When the host choose to start the match, the event \texttt{START} is 
+broadcasted too, making the clients start the dashboard activity.
+
+\begin{figure}[t]
+ \centering
+ \includegraphics[scale=0.42]{StartMatch}
+ \caption{Schema of a start of a match}
+ \label{fig:gbStartMatch}
+\end{figure}
+
+\subsection{Data transmission}
+
+We decided to transfer all the data in JSON format, encoding with base64 all the
+non-text content.
+Of all the transport protocols available in Bluetooth, we have chosen RFCOMM to 
+carry data between devices. First of all, RFCOMM has a behaviour similar to TCP 
+and it is recommend for data transfers~\cite{bisdikian01}. Secondly, Android 
+offers a Bluetooth API based on RFCOMM that gives the possibility to open 
+sockets and treat them as standard TCP connections.
+RFCOMM is usually referred as a serial port emulator, it is make on top of 
+the L2CAP protocol~\cite{bisdikian01}, and allows up to 60 simultaneous 
+connections between two Bluetooth devices~\cite{aneesh12}.
+
+The communications during the game creation are synchronous when 
+\texttt{MEMBER\_CONNECTED} and \texttt{CURRENT\_STATE} events happen, while 
+they are asynchronous for packets regarding \texttt{POKE} 
+and \texttt{MEMBER\_READY} actions. Instead, during the match, all the 
+communications are performed synchronously since the order of the transactions 
+is very important.
+
+\subsection{Known issues}
+
+During the app development we have faced some issues that we were able to 
+partially solve.
+
+\subsubsection{Image dimensions}
+
+As already said, images too large cause long time transmission with Bluetooth, 
+resulting in the application hanging while transferring data, providing bad 
+usability. In order to solve this problem, we have reduced image dimension, 
+scaling it, converting to the JPEG format and reducing its overall quality.
```

Comment:
Questa issue deve essere in parte risolta dalla compressione da implementare come suggerita dal prof. 

Resta invece valido il fatto che se tanti client si connettono l'host ha il Main Thread bloccato per troppo tempo.

Reply:
Allora dico che abbiamo parzialmente risolto inserendo lo scaling dell'immagine così argomento un po' di più

Reply:
Sì esatto, ho aggiunto lo scaling dell'immagine a 512x512, puoi dire che l'immagine può essere deformata  se di natura rettangolare